### ISSUE TITLE
Establish public API for DeploymentHandle

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle#2336]
For `PlayRun` and continuous build, we developed an internal deployment handle and deployment registry feature.  This is only used by `PlayRun`, but there may be existing plugins that use the internal APIs (like Ratpack's Gradle plugin).

We want to introduce public APIs for these internal types:
`org.gradle.deployment.internal.DeploymentHandle`
`org.gradle.deployment.internal.DeploymentRegistry`

There are a few issues that need to be addressed/determined before making this public:
- How would a task/plugin get access to the `DeploymentRegistry` or should this be kept internal? It would be nice if this could be some kind of thing that other tasks could depend on.
- `org.gradle.deployment.internal.DeploymentRegistry#onNewBuild()` should remain an internal API.
- We need to add `stop()` to the interface and not extend `Stoppable` (internal)
- The `id` for identifying a deployment is just a String, but there's no guarantees about how this is unique across projects/builds.  We should have something stronger for this.
- `org.gradle.deployment.internal.DeploymentRegistry#register()` should take a `Class<DeploymentHandle>` and Gradle should instantiate the handle.  e.g., `register(id, Class, args...)`
- `PlayRun` should be implemented on top of the new API.

To make this more generally useful, I think we should also do a few other things:
- We should change `PlayRun` to only start, but not wait for the application to exit when not in continuous build.
- When not a continuous build, we should list and wait for any currently running deployments after the build is finished.  We should allow either ctrl+D or ctrl+C to cancel them.  This would appear to be a "one shot" continuous build where we'd show "WAITING".
- Take a look at https://github.com/gradle/gradle/issues/2252

It's a common pattern that people want to be able to 
1) start some service 
2) do something else 
3) optionally, clean-up.

A deployment API would allow someone to start a "deployment" and provide enough information for Gradle to manage its lifecycle during the build and after it.  I would expect that we would provide some simple deployment handle implementations (like we do for `PlayRun`, but maybe a generic `Exec` and `JavaExec`-like backed services).

Interested in what you might think of this @ghale @adammurdoch

**Comment by adammurdoch** [comment]
I think this would be an excellent move.

**Comment by oehme** [comment]
> A deployment API would allow someone to start a "deployment" and provide enough information for Gradle to manage its lifecycle during the build and after it.

Another related issue: Integration tests often need something to be deployed and teared down after, but only if they actually run. Expressing this with tasks is clumsy, because the "deply" step will run even if the tests are up-to-date. It would be great if this deployment could be expressed as a task input instead.

**Comment by alkemist** [comment]
We should also consider the tooling implications here. _Ideally_, the “runnable” things in a build would be discoverable by tooling, and could be started/stopped in some manner.

Again, _ideally_, in my IDE I can see a list of “apps” that the can be developed with the current build. I want to then indicate that I now want to work on one or more and have them now continuously build. It's likely that there are also other useful variants on this workflow.

I don't want to kill this initiative by creeping the scope, but I do think we need to look a little bit further ahead here and lay out some kind of longer term direction than just opening up what we currently have.

**Comment by alkemist** [comment]
I think we should probably rename this card. We should be thinking about what functionality we want to enable, rather than what APIs to make public.

**Comment by mkobit** [comment]
Would definitely look forward to some sort of API around this to provide better integration with continuous mode.

**Comment by mkurz** [comment]
Hi,

Play maintainer here.

We are in the process of evaluating on how to best design a offical Gradle plugin for the Play Framework and are wondering if this issue can be discussed again, since it seems it would be helpful to have to described abilities in a public API.

Can this issue be re-opened?

Thanks!

**Comment by big-guy** [comment]
@mkurz I still think Gradle should provide this public API somehow, but we've struggled to find a long term maintainer for the [Play support we developed years ago](https://github.com/gradle/playframework).  

Is this something that the playframework organization would be interested in helping out with?

**Comment by mkurz** [comment]
@big-guy, actually, @ihostage is working on an official Play Gradle plugin that will reside in the Play repo and eventually supersede https://github.com/gradle/playframework (though it's inspired by it and will likely (?) reuse some of the work that was done there). The plan is for Play itself to offer a Gradle plugin that is fully integrated into our release process, tested, and documented. By having it in our Play repo, we can ensure it always works with the current Play version, instead of hoping that https://github.com/gradle/playframework catches up with the newest Play releases.

We are also already adjusting the Play codebase to work seamlessly with this new Gradle plugin by offering and adjusting APIs, etc., to avoid dirty hacks for the plugins to work. This is something that https://github.com/gradle/playframework wasn't able to do since it was a third-party project. Also, they basically never reached out to us, so they ended up using a lot of reflection, etc., to make things work (at least according to @ihostage, who knows that code very well and came up with writing our own plugin).

So, in short, we from Play, or better, at this point, @ihostage, would maintain an official Play Gradle plugin.

If you could help us here and collaborate with @ihostage to provide the necessary APIs, that would be great!

**Comment by mkurz** [comment]
Oh, and @ihostage is already actively working on that. This isn't just talk; he has delved deep into Gradle plugin development. As mentioned, he is highly familiar with the existing https://github.com/gradle/playframework plugin and has already developed a working prototype, though it still requires some refinement. Notably, he has prior experience having implemented the official Twirl Gradle Plugin, available here: https://plugins.gradle.org/plugin/org.playframework.twirl (Source: https://github.com/playframework/twirl/tree/main/gradle-twirl).

His goal is to complete the Play Gradle plugin by the end of this year 🤞.

**Comment by big-guy** [comment]
> His goal is to complete the Play Gradle plugin by the end of this year

That's great to hear. 

Once you get to a level of feature parity/stability that you're comfortable with supporting, let me know and we'll update gradle/playframework to warn and point people to the official plugin.

**Comment by big-guy** [comment]
> If you could help us here and collaborate with @ihostage to provide the necessary APIs, that would be great!

If there's anything outside of this issue, please open a new issue and I'll let the triaging team know to keep an eye out for them.

**Comment by pioterj** [comment]
@mkurz @ihostage This is great news and something we have been hoping to see for a long time. Having the maintainers of the Play Framework maintain the Gradle plugin too makes complete sense. We reached out to Lightbend some years ago, and they were not interested in taking the plugin over. 

Let us know if you need any support from our side. We can establish a more fine-grained communication channel, such as Slack, if needed, and/or set up a call to coordinate. In addition to the technical work, I think it would be a good idea to coordinate announcing this transition to the community. Let’s stay in touch.

**Comment by big-guy** [comment]
I wonder if BuildServices cover most of what this needs

**Comment by whyoleg** [comment]
Not sure if this's the best place for this kind of feedback, but such an API would also be a nice way for [Dokka](https://github.com/Kotlin/dokka) to let users see a live preview of their documentation while working on it by running a Gradle task.

Currently, the most straightforward way to preview Dokka documentation is to use IntelliJ's [built-in web server](https://www.jetbrains.com/help/phpstorm/php-built-in-web-server.html). 

So even now you could:
- run `:dokkaGenerateHtml -t`
- open the link provided by the Dokka task (e.g., `http://localhost:63342/PATH/build/dokka/html/index.html`) which points to IntelliJ's built-in web server
- check docs and update some kdocs
- the Dokka task will be automatically rerun by Gradle
- reload browser page and see updated docs

There are some problems with this approach:
- IntelliJ's built-in web server sometimes works incorrectly ([KT-79728](https://youtrack.jetbrains.com/issue/KT-79728))
- user might be using something other than IntelliJ.

That's why it would be nice if we could have a separate task for this (e.g. `:dokkaPreviewHtml`), which would start _some_ server to host files and support continuous builds.

Starting a server with JDK 18 could be a simple task:

```kotlin
import com.sun.net.httpserver.SimpleFileServer
import java.net.InetSocketAddress

// requires JDK 18: https://openjdk.org/jeps/408
tasks.register("dokkaPreview") {
    group = "dokka"
    val dokkaOutputDirectory = tasks.dokkaGeneratePublicationHtml.flatMap { it.outputDirectory }
    dependsOn(dokkaOutputDirectory)
    doLast {
        val server = SimpleFileServer.createFileServer(
            InetSocketAddress(0),
            dokkaOutputDirectory.get().asFile.toPath(),
            SimpleFileServer.OutputLevel.NONE
        )

        server.start()
        try {
            println("Preview Dokka HTML at http://localhost:${server.address.port}")
            Thread.sleep(Long.MAX_VALUE)
        } finally {
            server.stop(0)
        }
    }
}
```

But at this moment, it's not possible for this server to run when executing tasks with `-t`.
I've tried to implement this with `BuildService`, but it will not work too, because the service will be closed after the task using it is executed, since `BuildService` doesn't work across task reruns.

**Comment by ihostage** [comment]
@whyoleg, you can use the `DeploymentHandle` and `DeploymentRegistry` classes to implement the `dokkaPreview`, as you described. This should not be very difficult. 😉 

As example you can see a new Playframework Gradle plugin
1. [PlayRun](https://github.com/playframework/playframework/blob/4b97fa0af1652d8a5b10c12d0e2b419ea1e6ccf1/dev-mode/gradle-plugin/src/main/java/play/gradle/task/PlayRun.java) task as example for your `dokkaPreview` that use `DeploymentRegistry`.
2. [PlayRunHandle](https://github.com/playframework/playframework/blob/4b97fa0af1652d8a5b10c12d0e2b419ea1e6ccf1/dev-mode/gradle-plugin/src/main/java/play/gradle/internal/PlayRunHandle.java) (implements `DeploymentHandle`) as example how you can start and stop `SimpleFileServer` from your code.

**Comment by whyoleg** [comment]
> you can use the DeploymentHandle and DeploymentRegistry classes to implement the dokkaPreview, as you described. This should not be very difficult. 😉

@ihostage, as this is an `internal` Gradle API, we should be very careful with the task implementation to avoid breaking it in case the internal API is dropped/replaced. Or at least isolate the codepath so it's only used when the `dokkaPreview` task is used by the user. It's not impossible, it just needs resources, which we would rather spend now on other things. **But we accept PR contributions!**

The use case, though, is still relevant to mention so that it may be covered by the future stable API.
